### PR TITLE
Allow passing a 'translations' object in the configuration

### LINKED ISSUE #15713: [New installation methods] Allow passing a `translations` object in the configuration
The scope of this task is to add a new `translations` configuration object that has priority over the `CKEDITOR_TRANSLATIONS` object that the current translations register. See the [RFC for new installation methods](https://github.com/ckeditor/ckeditor5/issues/15502) for usage examples.

The `translations` object should accept the object (`translations: {}`) or an array of objects to be merged internally (`translations: [ {}, {} ]`).

We've already created a [working PoC](https://github.com/ckeditor/ckeditor5/pull/15346/files) that can be used, but it requires writing tests and a proper review process.

### PR DESCRIPTION
**Pull request by illia-stv**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (core,utils): Should use translations which passed in config. Closes #15713.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-core/src/editor/editor.ts [hunk 1]
```diff
@@ -277,8 +277,9 @@ export default abstract class Editor extends ObservableMixin() {
 
 		// Prefer the language passed as the argument to the constructor instead of the constructor's `defaultConfig`, if both are set.
 		const language = config.language || ( constructor.defaultConfig && constructor.defaultConfig.language );
+		const translations = config.translations || undefined;
```
Comment (arkflpc) [inline packages/ckeditor5-core/src/editor/editor.ts:280]: It is already `|| undefined`.

File: packages/ckeditor5-core/src/editor/editorconfig.ts [hunk 2]
```diff
@@ -533,6 +533,11 @@ export interface EditorConfig {
 	 * [order a trial](https://orders.ckeditor.com/trial/premium-features).
 	 */
 	licenseKey?: string;
+
+	/**
+	 * Translations to be used in editor.
+	 */
+	translations?: Translations | Array<Translations>;
```
Comment (arkflpc) [inline packages/ckeditor5-core/src/editor/editorconfig.ts:540]: Nitpick: there is `ArrayOrItem` helper type in `utils`.

File: packages/ckeditor5-utils/src/locale.ts [hunk 3]
```diff
@@ -107,13 +113,20 @@ export default class Locale {
 	 * @param options.contentLanguage The editor content language code in the
 	 * [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) format. If not specified, the same as `options.language`.
 	 * See {@link #contentLanguage}.
+	 * @param translations Translations passed as a editor config parameter.
 	 */
-	constructor( { uiLanguage = 'en', contentLanguage }: { readonly uiLanguage?: string; readonly contentLanguage?: string } = {} ) {
+	constructor( { uiLanguage = 'en', contentLanguage, translations }: { readonly uiLanguage?: string;
+			readonly contentLanguage?: string;
+			readonly translations?: Translations | Array<Translations> | undefined; } = {}
```
Comment (arkflpc) [inline packages/ckeditor5-utils/src/locale.ts:120]: Nitpick: `ArrayOrItem`.

Comment (arkflpc) [inline packages/ckeditor5-utils/src/locale.ts:120]:
```suggestion
	constructor( { uiLanguage = 'en', contentLanguage, translations }: {
			readonly uiLanguage?: string;
			readonly contentLanguage?: string;
			readonly translations?: Translations | Array<Translations> | undefined;
	} = {}
```

File: packages/ckeditor5-utils/src/locale.ts [hunk 4]
```diff
@@ -165,7 +178,11 @@ export default class Locale {
  * @param values A value or an array of values that will fill message placeholders.
  * For messages supporting plural forms the first value will determine the plural form.
  */
-export type LocaleTranslate = ( message: string | Message, values?: number | string | ReadonlyArray<number | string> ) => string;
+export type LocaleTranslate = (
+	message: string | Message,
+	values?: number | string | ReadonlyArray<number | string>,
+	translations?: { [key: string]: string } | null
```
Comment (arkflpc) [inline packages/ckeditor5-utils/src/locale.ts:184]: Why 3rd parameter? Seems unused. The function of this type is initialized in this file in line 130. It has 2 parameters only.

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 5]
```diff
@@ -9,16 +9,13 @@
  * @module utils/translation-service
  */
 
+import type { Translations, Editor } from '@ckeditor/ckeditor5-core';
```
Comment (arkflpc) [inline packages/ckeditor5-utils/src/translation-service.ts:12]: I'm worried about circular dependency. Do we really want to make `utils` dependent on `core`? I know it is `devDependencies`, but still.
@filipsobol, @niegowski ?

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 6]
```diff
@@ -114,18 +111,33 @@ if ( !global.window.CKEDITOR_TRANSLATIONS ) {
 export function add(
 	language: string,
 	translations: { readonly [ messageId: string ]: string | ReadonlyArray<string> },
-	getPluralForm?: ( n: number ) => number
+	getPluralForm?: ( n: number ) => number,
+	editor?: Editor
```
Comment (arkflpc) [inline packages/ckeditor5-utils/src/translation-service.ts:115]: What is the purpose?
`add` function was supposed to amend the global variable with new translations. `editor` will have the config already initialized with all the provided translations. How is `add` useful for configuring the specific instance of editor?

**Review by arkflpc — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-utils/src/locale.ts (hunk 4 above)
Comment (illia-stv) [reply packages/ckeditor5-utils/src/locale.ts:184]: It is used right here: https://github.com/ckeditor/ckeditor5/pull/15734/files#diff-bee38257de0d7815896e6ced52d1dca3534beb0e4f4faa84182cdf4dbdaf7ce6R127-R129

**Review by filipsobol — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 7]
```diff
@@ -177,17 +180,17 @@ export function _translate( language: string, message: Message, quantity: number
...
 	}
 
-	const numberOfLanguages = getNumberOfLanguages();
+	const numberOfLanguages = getNumberOfLanguages( translations );
 
 	if ( numberOfLanguages === 1 ) {
 		// Override the language to the only supported one.
 		// This can't be done in the `Locale` class, because the translations comes after the `Locale` class initialization.
-		language = Object.keys( global.window.CKEDITOR_TRANSLATIONS )[ 0 ];
+		language = translations ? Object.keys( translations )[ 0 ] : Object.keys( global.window.CKEDITOR_TRANSLATIONS )[ 0 ];
```
Comment (pszczesniak) [inline packages/ckeditor5-utils/src/translation-service.ts:188]: Maybe we can simplify this a bit:

```suggestion
		language = Object.keys( translations || global.window.CKEDITOR_TRANSLATIONS )[ 0 ];
```

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 8]
```diff
@@ -196,8 +199,9 @@ export function _translate( language: string, message: Message, quantity: number
 		return message.string;
 	}
 
-	const dictionary = global.window.CKEDITOR_TRANSLATIONS[ language ].dictionary;
-	const getPluralForm = global.window.CKEDITOR_TRANSLATIONS[ language ].getPluralForm || ( n => n === 1 ? 0 : 1 );
+	const dictionary = ( translations ? translations : global.window.CKEDITOR_TRANSLATIONS )[ language ].dictionary;
+	const getPluralForm = ( translations ? translations : global.window.CKEDITOR_TRANSLATIONS )[ language ].getPluralForm ||
```
Comment (pszczesniak) [inline packages/ckeditor5-utils/src/translation-service.ts:203]: This ternary conditional check is repeated twice:

```
translations ? translations : global.window.CKEDITOR\_TRANSLATIONS
```

Lets let's make it as a constant and use it.

**Review by pszczesniak — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-core/src/context.ts [hunk 9]
```diff
@@ -146,7 +146,13 @@ export default class Context {
 	 * @param config The context configuration.
 	 */
 	constructor( config?: ContextConfig ) {
-		this.config = new Config<ContextConfig>( config, ( this.constructor as typeof Context ).defaultConfig );
+		/**
+		 * We don't pass translations to the config, because its behavior of splitting keys
+		 * with dots (e.g. `resize.width` => `resize: { width }`) breaks the translations.
+		 */
+		const { translations, ...rest } = config || {};
```
[the remark this answers is not in the document]
Comment (filipsobol) [reply packages/ckeditor5-core/src/context.ts:153]: I'd also like to see a test that ensures that `translations` are properly passed to the `Config` and `Context` classes when the editor is created.

**Review by filipsobol — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-core/src/editor/editor.ts [hunk 10]
```diff
@@ -277,15 +277,16 @@ export default abstract class Editor extends ObservableMixin() {
...
-		this._context = config.context || new Context( { language } );
+		this._context = config.context || new Context( { language, translations } );
 		this._context._addEditor( this, !config.context );
 
 		// Clone the plugins to make sure that the plugin array will not be shared
 		// between editors and make the watchdog feature work correctly.
 		const availablePlugins = Array.from( constructor.builtinPlugins || [] );
 
-		this.config = new Config<EditorConfig>( config, constructor.defaultConfig );
+		this.config = new Config<EditorConfig>( { ...config, translations: undefined }, constructor.defaultConfig );
```
Comment (filipsobol) [inline packages/ckeditor5-core/src/editor/editor.ts:289]: Let's be consistent and use the same destructuring method as in `context.ts`.

**Review by filipsobol — COMMENTED** [review]
I added a comment asking for a small change. Otherwise the rest of the PR looks fine.

I tested these changes with new translations generated using Rollup, and a quick test shows that both pieces work well together.

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 11]
```diff
@@ -177,17 +180,18 @@ export function _translate( language: string, message: Message, quantity: number
 		throw new CKEditorError( 'translation-service-quantity-not-a-number', null, { quantity } );
 	}
 
-	const numberOfLanguages = getNumberOfLanguages();
+	const normalized: Translations = translations || global.window.CKEDITOR_TRANSLATIONS;
```
Comment (pszczesniak) [inline packages/ckeditor5-utils/src/translation-service.ts:183]: Let's make this variable name to be more descriptive, maybe `normalizedTranslations` would be better?

**Review by pszczesniak — CHANGES_REQUESTED** [review]
Just one small suggestion to change/update, other than that - PR looks fine :+1:

**Review by pszczesniak — APPROVED** [review]

File: packages/ckeditor5-core/src/editor/editor.ts [hunk 12]
```diff
@@ -278,14 +278,20 @@ export default abstract class Editor extends ObservableMixin() {
...
+
+		this._context = config.context || new Context( { language, translations } );
 		this._context._addEditor( this, !config.context );
 
 		// Clone the plugins to make sure that the plugin array will not be shared
 		// between editors and make the watchdog feature work correctly.
 		const availablePlugins = Array.from( constructor.builtinPlugins || [] );
 
-		this.config = new Config<EditorConfig>( config, constructor.defaultConfig );
+		this.config = new Config<EditorConfig>( rest, constructor.defaultConfig );
```
Comment (niegowski) [inline packages/ckeditor5-core/src/editor/editor.ts:294]: Changes in this file are not reflected in the tests.

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 13]
```diff
@@ -163,9 +160,15 @@ export function add(
...
  * @param quantity The number of elements for which a plural form should be picked from the target language dictionary.
+ * @param translations Translations passed in editor config, if not provided use the global `window.CKEDITOR_TRANSLATIONS`.
  * @returns Translated sentence.
  */
-export function _translate( language: string, message: Message, quantity: number = 1 ): string {
+export function _translate(
+	language: string,
+	message: Message,
+	quantity: number = 1,
+	translations?: Translations | undefined
```
Comment (niegowski) [inline packages/ckeditor5-utils/src/translation-service.ts:170]:
```suggestion
	translations?: Translations
```

File: packages/ckeditor5-utils/src/translation-service.ts [hunk 14]
```diff
@@ -216,21 +220,34 @@ export function _translate( language: string, message: Message, quantity: number
...
+}
+
+/**
+ * If array then merge objects which are inside otherwise return given object.
+ *
+ * @internal
+ * @param translations Translations passed in editor config.
+ */
+export function _unifyTranslations(
+	translations: Translations | Array<Translations> | undefined
```
Comment (niegowski) [inline packages/ckeditor5-utils/src/translation-service.ts:235]:
```suggestion
	translations?: ArrayOrItem<Translations>
```

File: packages/ckeditor5-utils/src/locale.ts [hunk 15]
```diff
@@ -107,13 +112,19 @@ export default class Locale {
...
 	 * See {@link #contentLanguage}.
+	 * @param translations Translations passed as a editor config parameter.
 	 */
-	constructor( { uiLanguage = 'en', contentLanguage }: { readonly uiLanguage?: string; readonly contentLanguage?: string } = {} ) {
+	constructor( { uiLanguage = 'en', contentLanguage, translations }: {
+		readonly uiLanguage?: string;
+		readonly contentLanguage?: string;
+		readonly translations?: Translations | Array<Translations>;
+	} = {}
+	) {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-utils/src/locale.ts:122]: Or maybe this way of formatting would be better:
```suggestion
	constructor( {
		uiLanguage = 'en',
		contentLanguage,
		translations 
	}: {
		readonly uiLanguage?: string;
		readonly contentLanguage?: string;
		readonly translations?: ReadonlyArrayOrItem<Translations>;
	} = {} ) {
```

**Comment by illia-stv** [comment]
Due reverts in history we decided to PR to new one here: https://github.com/ckeditor/ckeditor5/pull/15792

**Pull request closed without merging**